File: bKap4_1_DemoVectorAutoregressiveModel.m
% Script for estimating an autoregressive (AR) model of inflation in-sample.
% The dependent variable is the inflation target. The AR lags use a reporting
% (publication) lag r, so the most recent usable value is y(t-(r+1)):
%   regressors for y(t):  y(t-(r+1)), y(t-(r+2)), ..., y(t-(r+p))
% Built on the bKap4 VAR utilities (fEstVAR / fEstOptVAR), now extended with
% an 'iReportLag' option.

% Clear console
clear; clc; close all;

% Set path
sOldPath = path;
addpath('./Utils/');

%% Settings
sDataPath  = './DATA/Liedtke/US/';
iReportLag = 1;       % reporting/publication lag r (first usable AR lag = r+1)
iNumLags   = 1;       % number of AR lags p (used for the single fit below)
vLagGrid   = 1:12;    % candidate AR orders for AIC-based selection

%% Load data
% CSV layout: col 1 = observation_date, col 2 = target (inflation),
%             col 3..end = predictors (unused in the pure AR model).
tData    = readtable([sDataPath, 'aggregated.csv']);
dtDates  = tData{:,1};
if ~isdatetime(dtDates); dtDates = datetime(dtDates); end
vY       = tData{:,2};       % inflation target

%% Estimation
% AR(p) with reporting lag. fEstVAR removes rows made missing by the lagging
% (and the target's own NaNs) internally.
rModelAR = fEstVAR(vY, [], 'iNumLags', iNumLags, 'iReportLag', iReportLag, ...
    'lEstAlpha', true, 'lGetStats', true);

fprintf('AR(%d) on inflation with reporting lag r = %d\n', iNumLags, iReportLag);
fprintf('  Intercept        : %8.4f (t = %6.2f)\n', rModelAR.vAlpha, rModelAR.vAlphaT);
for k = 1:iNumLags
    fprintf('  Coef y(t-%-2d)      : %8.4f (t = %6.2f)\n', ...
        iReportLag + k, rModelAR.mPhi(k), rModelAR.mPhiT(k));
end
fprintf('  R2               : %8.4f\n', rModelAR.vR2);
fprintf('  AIC              : %8.2f\n', rModelAR.dAIC);

%% Select the AR order by AIC (reporting lag held fixed)
rModelOpt = fEstOptVAR(vY, [], 'vNumLags', vLagGrid, 'iReportLag', iReportLag);
fprintf('\nOptimal AR order by AIC over [%s]: p = %d  (reporting lag r = %d)\n', ...
    num2str(vLagGrid), rModelOpt.iNumLags, iReportLag);

% Restore path
path(sOldPath);

File: python/bKap4_1_DemoVectorAutoregressiveModel.py
"""
bKap4_1_DemoVectorAutoregressiveModel.py
Python port of bKap4_1_DemoVectorAutoregressiveModel.m.

In-sample AR model of inflation with a reporting (publication) lag r, so the
regressors for y(t) are y(t-(r+1)), ..., y(t-(r+p)). The AR order p is also
selected by AIC over a grid.

Run from the repo root:  python python/bKap4_1_DemoVectorAutoregressiveModel.py
"""

import numpy as np
import pandas as pd

import models

# =========================================================================
#  Settings
# =========================================================================
DATA_PATH  = "./DATA/Liedtke/US/aggregated.csv"
iReportLag = 1          # reporting lag r (first usable AR lag = r+1)
iNumLags   = 1          # number of AR lags p (single fit below)
vLagGrid   = range(1, 13)   # candidate AR orders for AIC selection

# =========================================================================
#  Load data
# =========================================================================
tData = pd.read_csv(DATA_PATH)
dtDates = pd.to_datetime(tData.iloc[:, 0])
vY = tData.iloc[:, 1].to_numpy(float)

# =========================================================================
#  Estimation
# =========================================================================
rModelAR = models.est_var(vY, None, num_lags=iNumLags, report_lag=iReportLag,
                          est_alpha=True, get_stats=True)

print(f"AR({iNumLags}) on inflation with reporting lag r = {iReportLag}")
print(f"  Intercept        : {rModelAR['alpha']:8.4f} (t = {rModelAR['alpha_t']:6.2f})")
for k in range(iNumLags):
    print(f"  Coef y(t-{iReportLag + k + 1:<2d})      : "
          f"{rModelAR['phi'][k]:8.4f} (t = {rModelAR['phi_t'][k]:6.2f})")
print(f"  R2               : {rModelAR['R2']:8.4f}")
print(f"  AIC              : {rModelAR['AIC']:8.2f}")

# =========================================================================
#  Select the AR order by AIC (reporting lag held fixed)
# =========================================================================
rModelOpt = models.est_opt_var(vY, None, num_lags_grid=vLagGrid,
                               report_lag=iReportLag)
print(f"\nOptimal AR order by AIC over [{' '.join(map(str, vLagGrid))}]: "
      f"p = {rModelOpt['num_lags']}  (reporting lag r = {iReportLag})")
